pytorch_tabnet/callbacks.py

```diff
@@ -0,0 +1,224 @@
+            callback.on_epoch_begin(epoch, logs)
+
+    def on_epoch_end(self, epoch, logs=None):
+        logs = logs or {}
+        for callback in self.callbacks:
+            callback.on_epoch_end(epoch, logs)
+
+    def on_batch_begin(self, batch, logs=None):
+        logs = logs or {}
+        for callback in self.callbacks:
+            callback.on_batch_begin(batch, logs)
+
+    def on_batch_end(self, batch, logs=None):
+        logs = logs or {}
+        for callback in self.callbacks:
+            callback.on_batch_end(batch, logs)
+
+    def on_train_begin(self, logs=None):
+        logs = logs or {}
+        logs["start_time"] = time.time()
+        for callback in self.callbacks:
+            callback.on_train_begin(logs)
+
+    def on_train_end(self, logs=None):
+        logs = logs or {}
+        for callback in self.callbacks:
+            callback.on_train_end(logs)
+
+
+@dataclass
+class EarlyStopping(Callback):
+    """EarlyStopping callback to exit the training loop if early_stopping_metric
+    does not improve by a certain amount for a certain
+    number of epochs.
+
+    Parameters
+    ---------
+    early_stopping_metric : str
+        Early stopping metric name
+    is_maximize : bool
+        Whether to maximize or not early_stopping_metric
+    tol : float
+        minimum change in monitored value to qualify as improvement.
+        This number should be positive.
+    patience : integer
+        number of epochs to wait for improvment before terminating.
+        the counter be reset after each improvment
+
+    """
+
+    early_stopping_metric: str
+    is_maximize: bool
+    tol: float = 0.0
+    patience: int = 5
+
+    def __post_init__(self):
+        self.best_epoch = 0
+        self.stopped_epoch = 0
+        self.wait = 0
+        self.best_weights = None
+        self.best_loss = np.inf
+        if self.is_maximize:
+            self.best_loss = -self.best_loss
+        super().__init__()
+
+    def on_epoch_end(self, epoch, logs=None):
+        current_loss = logs.get(self.early_stopping_metric)
+        if current_loss is None:
+            return
+        loss_change = current_loss - self.best_loss
+        max_improved = (self.is_maximize and loss_change > self.tol)
+        min_improved = ((not self.is_maximize) and (-loss_change > self.tol))
+        if max_improved or min_improved:
+            self.best_loss = current_loss
+            self.wait = 1
+            self.best_weights = copy.deepcopy(self.trainer.network.state_dict())
+            self.best_epoch = epoch
+        else:
+            if self.wait >= self.patience:
+                self.stopped_epoch = epoch + 1
+                self.trainer._stop_training = True
+            self.wait += 1
+
+    def on_train_end(self, logs=None):
+        self.trainer.best_epoch = self.best_epoch
+        self.trainer.best_cost = self.best_loss
+        self.trainer.network.load_state_dict(self.best_weights)
+        if self.stopped_epoch > 0:
```

Comment:
The msg occured only if the user ask for early stopping with a positive patience, if not, there is no early stopping.

![image](https://user-images.githubusercontent.com/56252443/95187538-91c08180-07cb-11eb-896f-27d3bd126753.png)

The actual msg is ok.
I delete `self.stopped_epoch` because the check can be done on `self.trainer._stop_training`

Reply:
I still don't understand what `self.stopped_epoch` represents and how it can be < 0.

The current message is not ok, because early stopping did not occur at epoch 1 but at epoch 3 as you can see. However best epoch was epoch 1 with best_valid_AUC etc...
That's what I meant in my previous message.

Reply:
Sorry, incomprehension, I fixed it

Reply:
no problem thank you!